cpp/src/arrow/compute/kernels/scalar_cast_nested.cc

```diff
@@ -150,6 +150,104 @@ void AddListCast(CastFunction* func) {
   DCHECK_OK(func->AddKernel(SrcType::type_id, std::move(kernel)));
 }
 
+struct CastStruct {
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    const CastOptions& options = CastState::Get(ctx);
+    const auto in_field_count =
+        checked_cast<const StructType&>(*batch[0].type()).num_fields();
```

Comment:
nit, but we could factor out the checked_cast of the types here and below

Reply:
Yea pretty sure that was what I tried, but let do that again tonight and post the traceback

Reply:
Cool this is working. My mistake was trying to do `const auto in_type = checked_cast<const StructType&>(*batch[0].type());` which yielded 

```
/arrow/cpp/src/arrow/type.h:972:20: error: use of deleted function ‘std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = arrow::StructType::Impl; _Dp = std::default_delete<arrow::StructType::Impl>]’
```

Not using auto here works well